app/src/docs/hooks/useLunr.js

```diff
@@ -0,0 +1,119 @@
+// @flow
+
+import { useState, useEffect, useMemo } from 'react'
+import lunr from 'lunr'
+
+const InvalidIndexError = Error('Lunr index could not be parsed. Check that your index exists and is valid.')
+const InvalidStoreError = Error('Lunr store could not be parsed. Check that your store exists and is valid.')
+
+const useLunr = (query: string, providedIndex: any, providedStore: any) => {
+    const [index, setIndex] = useState(null)
+    const [store, setStore] = useState(null)
+
+    useEffect(() => {
+        const processedIndex =
+            typeof providedIndex === 'string'
+                ? lunr.Index.load(JSON.parse(providedIndex))
+                : providedIndex
+
+        if (!processedIndex) {
+            throw InvalidIndexError
+        }
+
+        setIndex(processedIndex)
+    }, [providedIndex])
+
+    useEffect(() => {
+        const processedStore =
+            typeof providedStore === 'string'
+                ? JSON.parse(providedStore)
+                : providedStore
+
+        if (!processedStore) {
+            throw InvalidStoreError
+        }
+
+        setStore(processedStore)
+    }, [providedStore])
```

Comment:
So the way the state and these effects are set up will cause the useMemo to re-run once for each parameter that's updated.

so on the first run, it'll do something like:

1. run the memo without either index or store
2. run the index processing effect. setIndex in the effect queues a re-render.
3. re-run the memo with the newly processed index
4. run the store processing effect. setStore in the effect queues a re-render.
5. re-run the memo with the newly processed store

I'm thinking the effects and state could go away entirely and instead `useMemo` like this:

```js
const index = useMemo(() => {
    const processedStore =
            typeof providedStore === 'string'
                ? JSON.parse(providedStore)
                : providedStore

        if (!processedStore) {
            throw InvalidStoreError()
        }
        return processedStore
}, [providedStore])

const index = useMemo(() => {
        const processedIndex =
            typeof providedIndex === 'string'
                ? lunr.Index.load(JSON.parse(providedIndex))
                : providedIndex

        if (!processedIndex) {
            throw InvalidIndexError()
        }
        return processedIndex
}, [providedIndex])
// etc
```

This should have the following benefits:
* both the store and index will always be updated (if necessary) in the same render, instead of staggered and possibly using stale data.
* they will be available on the first render when all parameters are set
* they don't enqueue any additional renders

Reply:
@timoxley I've tidied this up now so it works a bit differently - cast another eye if you get bored!